File: nic/apollo/agent/core/mirror.cc
//------------------------------------------------------------------------------
// {C} Copyright 2019 Pensando Systems Inc. All rights reserved
// -----------------------------------------------------------------------------

#include "nic/apollo/agent/core/state.hpp"
#include "nic/apollo/agent/trace.hpp"
#include "nic/apollo/agent/core/mirror.hpp"

namespace core {

sdk_ret_t
mirror_session_create (pds_obj_key_t *key,
                       pds_mirror_session_spec_t *spec,
                       pds_batch_ctxt_t bctxt)
{
    sdk_ret_t ret;

    if (!agent_state::state()->pds_mock_mode()) {
        if ((ret = pds_mirror_session_create(spec, bctxt)) != SDK_RET_OK) {
            PDS_TRACE_ERR("Failed to create mirror session {}, err {}",
                          spec->key.id, ret);
            return ret;
        }
    }
    return SDK_RET_OK;
}

sdk_ret_t
mirror_session_update (pds_obj_key_t *key,
                       pds_mirror_session_spec_t *spec,
                       pds_batch_ctxt_t bctxt)
{
    sdk_ret_t ret;

    if (!agent_state::state()->pds_mock_mode()) {
        if ((ret = pds_mirror_session_update(spec, bctxt)) != SDK_RET_OK) {
            PDS_TRACE_ERR("Failed to update mirror session {}, err {}",
                          spec->key.id, ret);
            return ret;
        }
    }
    return SDK_RET_OK;
}

sdk_ret_t
mirror_session_delete (pds_obj_key_t *key, pds_batch_ctxt_t bctxt)
{
    sdk_ret_t ret;

    if (!agent_state::state()->pds_mock_mode()) {
        if ((ret = pds_mirror_session_delete(key, bctxt)) != SDK_RET_OK) {
            PDS_TRACE_ERR("Failed to mirror session nh {}, err {}",
                          key->str(), ret);
            return ret;
        }
    }
    return SDK_RET_OK;
}

sdk_ret_t
mirror_session_get (pds_obj_key_t *key, pds_mirror_session_info_t *info)
{
    sdk_ret_t ret = SDK_RET_OK;

    if (!agent_state::state()->pds_mock_mode()) {
        ret = pds_mirror_session_read(key, info);
    } else {
        memset(&info->stats, 0, sizeof(info->stats));
        memset(&info->status, 0, sizeof(info->status));
    }
    return ret;
}

sdk_ret_t
mirror_session_get_all (mirror_session_read_cb_t mirror_session_read_cb,
                        void *ctxt)
{
    return pds_mirror_session_read_all(mirror_session_read_cb, ctxt);
}

}    // namespace core

File: nic/agent/dscagent/pipeline/apulu/mirror.go
// {C} Copyright 2020 Pensando Systems Inc. All rights reserved.

// +build apulu

package apulu

import (
	"context"
	"fmt"
	"net"

	"github.com/pkg/errors"

	"github.com/pensando/sw/nic/agent/dscagent/pipeline/apulu/utils"
	"github.com/pensando/sw/nic/agent/dscagent/types"
	"github.com/pensando/sw/nic/agent/protos/netproto"
	halapi "github.com/pensando/sw/nic/apollo/agent/gen/pds"
	"github.com/pensando/sw/venice/utils/log"
)

// HandleMirrorSession handles crud operations on mirror session
func HandleMirrorSession(infraAPI types.InfraAPI, client halapi.MirrorSvcClient, oper types.Operation, mirror netproto.MirrorSession,
	vpcID uint64) error {
	switch oper {
	case types.Create:
		return createMirrorSessionHandler(infraAPI, client, mirror, vpcID)
	case types.Update:
		return updateMirrorSessionHandler(infraAPI, client, mirror, vpcID)
	case types.Delete:
		return deleteMirrorSessionHandler(infraAPI, client, mirror)
	default:
		return errors.Wrapf(types.ErrUnsupportedOp, "Op: %s", oper)
	}
}

func createMirrorSessionHandler(infraAPI types.InfraAPI, client halapi.MirrorSvcClient, mirror netproto.MirrorSession, vpcID uint64) error {
	mirrorReq := convertMirrorSession(infraAPI, mirror, vpcID)
	resp, err := client.MirrorSessionCreate(context.Background(), mirrorReq)
	if resp != nil {
		if err := utils.HandleErr(types.Create, resp.ApiStatus, err, fmt.Sprintf("Create Failed for %s | %s", mirror.GetKind(), mirror.GetKey())); err != nil {
			return err
		}
	}

	dat, _ := mirror.Marshal()
	if err := infraAPI.Store(mirror.Kind, mirror.GetKey(), dat); err != nil {
		log.Error(errors.Wrapf(types.ErrBoltDBStoreCreate, "MirrorSession: %s | Err: %v", mirror.GetKey(), err))
		return errors.Wrapf(types.ErrBoltDBStoreCreate, "MirrorSession: %s | Err: %v", mirror.GetKey(), err)
	}
	return nil
}

func updateMirrorSessionHandler(infraAPI types.InfraAPI, client halapi.MirrorSvcClient, mirror netproto.MirrorSession, vpcID uint64) error {
	mirrorReq := convertMirrorSession(infraAPI, mirror, vpcID)
	resp, err := client.MirrorSessionUpdate(context.Background(), mirrorReq)
	if resp != nil {
		if err := utils.HandleErr(types.Create, resp.ApiStatus, err, fmt.Sprintf("Create Failed for %s | %s", mirror.GetKind(), mirror.GetKey())); err != nil {
			return err
		}
	}
	dat, _ := mirror.Marshal()

	if err := infraAPI.Store(mirror.Kind, mirror.GetKey(), dat); err != nil {
		log.Error(errors.Wrapf(types.ErrBoltDBStoreUpdate, "MirrorSession: %s | Err: %v", mirror.GetKey(), err))
		return errors.Wrapf(types.ErrBoltDBStoreUpdate, "MirrorSession: %s | Err: %v", mirror.GetKey(), err)
	}
	return nil
}

func deleteMirrorSessionHandler(infraAPI types.InfraAPI, client halapi.MirrorSvcClient, mirror netproto.MirrorSession) error {
	mirrorDelReq := &halapi.MirrorSessionDeleteRequest{
		Id: []uint32{
			uint32(mirror.Status.MirrorSessionID),
		},
	}

	resp, err := client.MirrorSessionDelete(context.Background(), mirrorDelReq)
	if resp != nil {
		if err := utils.HandleErr(types.Delete, resp.ApiStatus[0], err, fmt.Sprintf("MirrorSession: %s", mirror.GetKey())); err != nil {
			return err
		}
	}

	if err := infraAPI.Delete(mirror.Kind, mirror.GetKey()); err != nil {
		log.Error(errors.Wrapf(types.ErrBoltDBStoreDelete, "MirrorSession: %s | Err: %v", mirror.GetKey(), err))
		return errors.Wrapf(types.ErrBoltDBStoreDelete, "MirrorSession: %s | Err: %v", mirror.GetKey(), err)
	}
	return nil
}

func convertMirrorSession(infraAPI types.InfraAPI, mirror netproto.MirrorSession, vpcID uint64) *halapi.MirrorSessionRequest {
	var mirrorSpecs []*halapi.MirrorSessionSpec
	mgmtIP, _, _ := net.ParseCIDR(infraAPI.GetConfig().MgmtIP)
	for _, c := range mirror.Spec.Collectors {
		m := &halapi.MirrorSessionSpec{
			Id:      uint32(mirror.Status.MirrorSessionID),
			SnapLen: mirror.Spec.PacketSize,
			Mirrordst: &halapi.MirrorSessionSpec_ErspanSpec{
				ErspanSpec: &halapi.ERSpanSpec{
					VPCId: utils.ConvertID64(vpcID)[0],
					Erspandst: &halapi.ERSpanSpec_DstIP{
						DstIP: utils.ConvertIPAddresses(c.ExportCfg.Destination)[0],
					},
					SrcIP:  utils.ConvertIPAddresses(mgmtIP.String())[0],
					SpanId: uint32(mirror.Status.MirrorSessionID),
				},
			},
		}
		mirrorSpecs = append(mirrorSpecs, m)
	}
	return &halapi.MirrorSessionRequest{
		Request: mirrorSpecs,
	}
}
